### PR TITLE
Fix Issue #659

### PR DESCRIPTION
**Pull request by Hadis-Fard**
issue #659 : socket is assigned and attempted to connect after the connection has transitioned to final state by connect timeout.

### CODE REVIEW

File: src/connection.js [hunk 1]
```diff
@@ -689,17 +689,22 @@ class Connection extends EventEmitter {
       if (err) {
         return this.socketError(err);
       }
+      if (this.state === this.STATE.FINAL) {
+        if (socket) {
```
Comment (arthurschreiber) [inline src/connection.js]: I'm pretty sure `socket` is always set here.

Can we change the code to do an early return instead?

```js
if (this.state === this.STATE.FINAL) {
  socket.destroy();
  return;
}

this.socket = socket;
...
```

Also, could you add an empty newline before the `if` statement here?

**Comment by arthurschreiber** [comment]
This is great, thank you very much! I left a tiny comment regarding the code style, other than that, this is looking great! 👍

Can we also add an integration test for this, somehow?

File: src/connection.js [hunk 2]
```diff
@@ -690,16 +690,21 @@ class Connection extends EventEmitter {
...
-      this.socket.on('close', this.socketClose);
-      this.socket.on('end', this.socketEnd);
-      this.messageIo = new MessageIO(this.socket, this.config.options.packetSize, this.debug);
-      this.messageIo.on('data', (data) => { this.dispatchEvent('data', data); });
-      this.messageIo.on('message', () => { this.dispatchEvent('message'); });
-      this.messageIo.on('secure', this.emit.bind(this, 'secure'));
-
-      this.socketConnect();
+      if (this.state === this.STATE.FINAL) {
+          socket.destroy();
```
Comment (arthurschreiber) [inline src/connection.js]: The whitespace is still a bit off here. 😅

**Comment by Hadis-Fard** [comment]
@arthurschreiber the pause-resume test has failed on Appveyor build. The test is flaky and fails when I test on a remote server, there's also a discussion about it in [518](https://github.com/tediousjs/tedious/pull/518#issuecomment-342658545), could you please fix it?

**Comment by arthurschreiber** [comment]
Perfect! Thanks for following up with the style changes. 😍 I'll merge and release this ASAP!

**Comment by arthurschreiber** [comment]
This was just released as part of `tedious@2.2.4` (currently still tagged as `next`).

**Comment by pbarbiero** [comment]
@arthurschreiber is there any chance we can get this back-ported to the 1.x branch for those of us stuck for the time being?  We've started having servers die due to this recently, and we are not in a position to upgrade to v2 currently in some cases.

**Comment by arthurschreiber** [comment]
If I remember correctly, the upgrade to v2 should be pretty painless. The only real change was dropping support for node versions earlier to 4.

If possible I'd like to avoid spending time on backporting changes, but if you absolutely can not upgrade, I can go and do it. Please let me know! 😄

**Comment by pbarbiero** [comment]
Thanks Arthur,

If thats the main churn to v2, I can probably force it through our release process. I don't want to burden you with more releases in light of that.

I appreciate the response!

**Pull request merged**